In [1]:
import pandas as pd
import numpy as np
#imported 2 libraries called the numpy and the pandas

This is the data being used for the peer graded assessment.
The dataset revolves around the compressive strength of various concrete samples, depending on the quantities of diverse components employed in their creation. These components encompass:

1) Cement

2) Blast Furnace Slag

3) Fly Ash

4) Water

5) Superplasticizer

6) Coarse Aggregate

7) Fine Aggregate

In [3]:
#Reading the data through the data dataframe 
concrete_data = pd.read_csv('concrete_data.csv')
concrete_data.head()  #printing the first five rows of the dataset.

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

So here we know that the first concrete sample has 540 cubic meter of cement, zero blast furnace slag, zero fly ash, 162 cubic meter of water,2.5 cubic meter of superplatsicizer, 1040 cubic meter of coarse aggregate, 676 cubic meter of fine aggregate, 28 is the age and 79.99 is the strength.

In [4]:
#Checking the total number of data points 
concrete_data.shape

(1030, 9)

In [5]:
#Checking data for any missing values 
concrete_data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

In [6]:
# Removing all null data from the dataset 
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

As the data has no null values so now we can use this data to train our model.

For creating a Regression or a classification model we always need to divide the columns into two called the predictors and the target.

In [7]:
# Our target variable is Strength and others columns are the predictors
concrete_data_columns = concrete_data.columns
predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # Every columns except Strength
target = concrete_data['Strength']  # Strength column

Printing the predictors and the targets just to check the output

In [10]:
target.head()  #As we have only 1 target variable so one column gets printed

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

In [11]:
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

In [12]:
#Checking the number of predictors 
num_cols = predictors.shape[1]
num_cols

8

Moving forward we import keras and start building our regression model

In [16]:
import keras     #Keras is always imported from the backend tensorflow

In [18]:
from keras.models import Sequential    #We are importing the sequential layer from keras which helps us to create a model layer by layer in a linear stack
from keras.layers import Dense    #A dense layer is imported from layers to form a fully connected neural network layer

Creating the regression model

In [29]:
def regression_model():    #Created a function to create a regression model.
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(num_cols,)))   #This contains 10 nodes with the relu activation function(1st hidden layer)
    model.add(Dense(1))  #Ouput layer

    model.compile(optimizer='adam', loss='mean_squared_error')  #Compiling the model with the optimizer adam and keeping the loss as MSE.
    return model

Now we want to split the data so we will have to import sklearn to train and test the data

In [30]:
from sklearn.model_selection import train_test_split

Splitting the data into 70 % training and 30% testing data

In [31]:
X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=42)

Now we have to build our model so we call the regression function here that we created above

In [32]:
#Calling our model here
model = regression_model()


In [35]:
#Training our model with epochs=50
epochs = 50
model.fit(X_train, y_train, epochs=epochs, verbose=1)

Epoch 1/50
721/721 [==============================] - 0s 45us/step - loss: 119.0285
Epoch 2/50
721/721 [==============================] - 0s 74us/step - loss: 117.5786
Epoch 3/50
721/721 [==============================] - 0s 53us/step - loss: 115.0128
Epoch 4/50
721/721 [==============================] - 0s 56us/step - loss: 113.5443
Epoch 5/50
721/721 [==============================] - 0s 56us/step - loss: 112.1446
Epoch 6/50
721/721 [==============================] - 0s 76us/step - loss: 110.3606
Epoch 7/50
721/721 [==============================] - 0s 82us/step - loss: 110.3640
Epoch 8/50
721/721 [==============================] - 0s 54us/step - loss: 108.1255
Epoch 9/50
721/721 [==============================] - 0s 61us/step - loss: 106.8686
Epoch 10/50
721/721 [==============================] - 0s 62us/step - loss: 106.9790
Epoch 11/50
721/721 [==============================] - 0s 52us/step - loss: 104.4616
Epoch 12/50
721/721 [==============================] - 0s 65us/step - loss

Now we evaluate the metrics for the model formed 

In [36]:
loss_val = model.evaluate(X_test, y_test)
y_pred = model.predict(X_test)
loss_val

309/309 [==============================] - 0s 54us/step


88.5624806178812

As per the next step we need to compute the mean squared error

In [37]:
from sklearn.metrics import mean_squared_error

In [ ]:
#Printing the meand and standard deviation of the test data 
mean_square_error = mean_squared_error(y_test, y_pred)
mean = np.mean(mean_square_error)
standard_deviation = np.std(mean_square_error)
print(mean, standard_deviation)

Creating a list of 50 mean squared errors.

In the code below it iterates through a specified number of training and evaluation cycles (total_mean_squared_errors). In each cycle, the dataset is split into training and testing sets, and a neural network model is trained for a fixed number of epochs. The mean squared error (MSE) is then calculated for the model's predictions on the test set, and both the MSE and the mean squared error between true and predicted values are recorded. After the specified number of cycles, the code calculates and prints the mean and standard deviation of the collected mean squared errors, providing insights into the model's overall performance and its variability across multiple runs.

In [40]:
total_mean_squared_errors = 50
#The total number of times the model will be trained and evaluated.
epochs = 50
mean_squared_errors = []
#An empty list to store the mean squared errors for each run.

#
for i in range(0, total_mean_squared_errors):
    X_train, X_test, y_train, y_test = train_test_split(predictors, target, test_size=0.3, random_state=i)
    model.fit(X_train, y_train, epochs=epochs, verbose=0)
    MSE = model.evaluate(X_test, y_test, verbose=0)
    print("MSE "+str(i+1)+": "+str(MSE))
    y_pred = model.predict(X_test)
    mean_square_error = mean_squared_error(y_test, y_pred)
    mean_squared_errors.append(mean_square_error)

#Convert the list of mean squared errors to a NumPy array for easier calculation.
#Calculate the mean and standard deviation of the mean squared errors.
mean_squared_errors = np.array(mean_squared_errors)
mean = np.mean(mean_squared_errors)
standard_deviation = np.std(mean_squared_errors)

print('\n')
print("Printing the  mean and standard deviation of " +str(total_mean_squared_errors) + " mean squared errors without normalized data. Total number of epochs for each training is: " +str(epochs) + "\n")
print("Mean: "+str(mean))
print("Standard Deviation: "+str(standard_deviation))

MSE 1: 60.95088588381277
MSE 2: 59.251511434906895
MSE 3: 47.29962247089275
MSE 4: 55.29933651674141
MSE 5: 47.79275491708305
MSE 6: 63.485785968866935
MSE 7: 60.85775623506713
MSE 8: 52.96166507017265
MSE 9: 52.03372523236815
MSE 10: 52.94406905683499
MSE 11: 48.641539329849785
MSE 12: 49.58736862256689
MSE 13: 64.00039180274149
MSE 14: 51.817096500335005
MSE 15: 50.96957493754267
MSE 16: 45.11294789607471
MSE 17: 67.65628194963276
MSE 18: 58.7444251470967
MSE 19: 44.14449104136248
MSE 20: 48.52295504573094
MSE 21: 47.21819172090697
MSE 22: 47.96904576866372
MSE 23: 53.69606167444519
MSE 24: 46.77917398989779
MSE 25: 49.72985276898134
MSE 26: 49.42392861573056
MSE 27: 55.66692870797463
MSE 28: 53.03581790862346
MSE 29: 52.64100558086506
MSE 30: 48.384918916572644
MSE 31: 54.45677608502336
MSE 32: 65.0235270651413
MSE 33: 52.280843098958336
MSE 34: 50.84068100815066
MSE 35: 45.83617087552462
MSE 36: 60.85918259852141
MSE 37: 53.580956881871884
MSE 38: 51.762874739069765
MSE 39: 48.6207